corporate/lib/stripe.py

```diff
@@ -604,17 +608,164 @@ def setup_upgrade_checkout_session_and_payment_intent(
+
+                realm = new_plan.customer.realm
+                assert realm is not None
+
+                RealmAuditLog.objects.create(
+                    realm=realm,
+                    event_time=event_time,
+                    event_type=RealmAuditLog.CUSTOMER_SWITCHED_FROM_MONTHLY_TO_ANNUAL_PLAN,
+                    extra_data={
+                        "monthly_plan_id": plan.id,
+                        "annual_plan_id": new_plan.id,
+                    },
+                )
+                return new_plan, new_plan_ledger_entry
+
+            if plan.status == CustomerPlan.SWITCH_NOW_FROM_STANDARD_TO_PLUS:
+                standard_plan = plan
+                standard_plan.end_date = next_billing_cycle
+                standard_plan.status = CustomerPlan.ENDED
+                standard_plan.save(update_fields=["status", "end_date"])
+
+                (_, _, _, plus_plan_price_per_license) = compute_plan_parameters(
+                    CustomerPlan.PLUS,
+                    standard_plan.automanage_licenses,
+                    standard_plan.billing_schedule,
+                    standard_plan.customer.default_discount,
+                )
+                plus_plan_billing_cycle_anchor = standard_plan.end_date.replace(microsecond=0)
+
+                plus_plan = CustomerPlan.objects.create(
+                    customer=standard_plan.customer,
+                    status=CustomerPlan.ACTIVE,
+                    automanage_licenses=standard_plan.automanage_licenses,
+                    charge_automatically=standard_plan.charge_automatically,
+                    price_per_license=plus_plan_price_per_license,
+                    discount=standard_plan.customer.default_discount,
+                    billing_schedule=standard_plan.billing_schedule,
+                    tier=CustomerPlan.PLUS,
+                    billing_cycle_anchor=plus_plan_billing_cycle_anchor,
+                    invoicing_status=CustomerPlan.INITIAL_INVOICE_TO_BE_SENT,
+                    next_invoice_date=plus_plan_billing_cycle_anchor,
+                )
+
+                standard_plan_last_ledger = (
+                    LicenseLedger.objects.filter(plan=standard_plan).order_by("id").last()
+                )
+                assert standard_plan_last_ledger is not None
+                licenses_for_plus_plan = standard_plan_last_ledger.licenses_at_next_renewal
+                assert licenses_for_plus_plan is not None
+                plus_plan_ledger_entry = LicenseLedger.objects.create(
+                    plan=plus_plan,
+                    is_renewal=True,
+                    event_time=plus_plan_billing_cycle_anchor,
+                    licenses=licenses_for_plus_plan,
+                    licenses_at_next_renewal=licenses_for_plus_plan,
+                )
+                return plus_plan, plus_plan_ledger_entry
+
+            if plan.status == CustomerPlan.DOWNGRADE_AT_END_OF_CYCLE:
+                process_downgrade(plan)
```

Comment:
Just a small note that for this function to work fully, we'll need to move `process_downgrade` and any other functions this calls to `BillingSession` too.

Reply:
Done